library/md.c

```diff
@@ -455,9 +455,11 @@ int mbedtls_md_setup(mbedtls_md_context_t *ctx, const mbedtls_md_info_t *md_info
 
 int mbedtls_md_starts(mbedtls_md_context_t *ctx)
 {
+#if defined(MBEDTLS_MD_C)
     if (ctx == NULL || ctx->md_info == NULL) {
```

Comment:
Note that according to the work we did removing `NULL` pointer checks (#4313) the checks like this could probably be just

```
if (ctx->md_info == NULL) {
```

@mpg Thoughts?

Reply:
Before `CHECK_PARAMS`, since the PolarSSL days, high-level modules accepted null pointers for pointer-to-context arguments in many functions, making the function return a `BAD_INPUT_DATA` error. For md, this dates back to [the commit that created the module 1737385e04ef392927646ed939000436f2da2240](https://github.com/Mbed-TLS/mbedtls/commit/1737385e04ef392927646ed939000436f2da2240).

This is not documented for most functions (for example, in md, it's only documented for `mbedtls_md_info_from_ctx`), so by our own rules we could remove these null checks. But it's old established practice, so it's plausible that users rely on that behavior. If we remove them in the full-md case, I think this warrants a changelog entry.

Disclaimer: I am an opponent of this kind of null checks: they have a very limited benefit (they only make a difference on platforms where a null dereference at runtime doesn't crash the program, which is generally platforms where code size matters most — or with whole-program optimizing compilers, but the cases where such optimizations could trigger are very unlikely in practice, and likely to be discovered in basic testing).

(Just so we don't digress, this is solely about contexts, not about info pointers for which the null checks are meaningful.)

Reply:
> You're both right that according to our guidelines we should keep the `ctx->md_info == NULL` part of the check.

Actually I'm not entirely convinced about that. We usually don't do NULL checks for the reasons mentioned by Gilles. Here in `mbedtls_md_starts()` the point of checking if `ctx->md_info` is `NULL` is basically to ensure that `mbedtls_md_setup()` has been called beforehand (as the document says it must), which makes sense - we don't do NULL check but we do state checks, and in that case that's the same.

But now if we look at `mbedtls_md_update()` I'd say the check for `ctx->md_info == NULL` is not so meaningful: it ensures that `md_setup()` has been called, but not that `md_starts()` has been called. If we're doing state checks, we should do it for real, with a state enum (like for example in `chachapoly.c`). In `mbedtls_md_update()` and `mbedtls_md_finish()`, checking if `ctx->md_info == NULL` is a half-baked state check: we're just checking that the state is not so bad as to cause us to crash, but we're not checking it's good enough to give meaningful results (failing to call `starts()` results in invalid output). (Similar considerations would apply for the streaming `mbedtls_md_hmac()` API btw.)

Now, adding proper state checks in MD is clearly not in scope for the code size EPIC, so I guess it makes sense for now to just remove the `ctx == NULL` checks when `MD_C` is undefined, and keep the `ctx->md_info` checks as they are. But I don't think that's the way it should be ideally - more of a mix between how it's been so far and how it should be.

Reply:
I think regarding the `ctx == NULL` checks we're in agreement that they shouldn't be there, but we might differ on how to act on that opinion.
* Tom C seemed to suggest we remove them unconditionally now.
* Gilles and I think this wouldn't be ideal for compat reasons so let's only remove them when `!MD_C` for now.
* I think we should remove them unconditionally in 4.0 too.

Reply:
Regarding the `md_info` null checks: indeed in the case of `update` this check may not be the right one. I was just reminding the fact that many of the `info == NULL` checks are legitimate, because that's how we carry around the fact that the application tried to use an algorithm that isn't supported. For example
```
info = mbedtls_md_info_from_type(type);
ret = mbedtls_md_setup(&ctx, info);
```
is the expected way to attempt to use an algorithm that may or may not be supported. If `type` is not supported, `info` is null and `setup` reliably returns an error. On the other hand, `mbedtls_md_setup(NULL, ...)` is a programmer error and does not need to be handled reliably.

Reply:
Regarding state checks in md: given that the future is PSA, I don't think it's worth spending much time making the md/cipher API more robust against programmer errors.